bayesian/utility/v2/sa_response_builder.py

```diff
@@ -0,0 +1,69 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Author: Dharmendra G Patel <dhpatel@redhat.com>
+#
+"""Stack analyses API v2 response builder class."""
+
+
+class StackAnalysesResponseBuilder:
+    """Stack analysis response builder for API v2."""
+
+    def __init__(self, external_request_id, stack_result, recm_data):
+        """Response Builder, Build Json Response for Stack Analyses."""
+        self.external_request_id = external_request_id
+        self.stack_result = stack_result
+        self.recm_data = recm_data
+
+    def get_response(self):
+        """Aggregate, build and return json response for the given request id."""
+        stack_task_result = None
+        stack_audit = None
+        recommendations = []
+
+        if self.stack_result is not None and "task_result" in self.stack_result:
+            stack_task_result = self.stack_result.get("task_result", None)
+            stack_audit = stack_task_result.get("_audit", {})
+
+        if self.recm_data is not None and "task_result" in self.recm_data:
+            recommendations = self.recm_data.get("task_result", {}).get("recommendations", [])[0]
+
+        response_data = {}
+        if stack_task_result is not None:
+            response_data = {
+                "version": stack_audit.get("version", None),
+                "started_at": stack_audit.get("started_at", None),
+                "ended_at": stack_audit.get("ended_at", None),
+                "external_request_id": self.external_request_id,
+                "registration_status": stack_task_result.get("registration_status", ""),
+                "manifest_file_path": stack_task_result.get("manifest_file_path", ""),
+                "manifest_name": stack_task_result.get("manifest_name", ""),
+                "ecosystem": stack_task_result.get("ecosystem", ""),
+                "unknown_dependencies": stack_task_result.get("unknown_dependencies", ""),
+                "license_analysis": stack_task_result.get("license_analysis", ""),
+                "recommendation": {
+                    "companion": recommendations.get("companion", []),
+                    "manifest_file_path": recommendations.get("manifest_file_path", ""),
+                    "usage_outliers": recommendations.get("usage_outliers", [])
+                },
+                "registration_link": stack_task_result.get("registration_link", ""),
+                "analyzed_dependencies": stack_task_result.get("analyzed_dependencies", [])
```

Comment:
You may need to add only one field that is `recommendation`, no need to explicitly populate rest of the fields because it is already part of the `stack_task_result`

Reply:
I differ here a bit, I feel its good to keep output format as per API spec from this end point and create a by-pass by just coping entire stack_result as it to response. Say in future by any chance we have modified API for backbone / it gives wrong fields, same will not be passed on to client causing more havoc. It will be filtered by this fixed response structure. All I am thinking is to keep output structure of SA V2 as per Spec irrespective of what response I get from backbone. Let me know if you need a discussion on this.

Reply:
You are always going to depend on what backbone persists. If backbone changes what it persists, then there are no changes required here.  I think this layer should only serve what backbone persists.

Reply:
This is not about persist data, API response is about format of the response. I think format of the response should not change, it might create unexpected parsing errors in client. Response format is what defined by API spec, data in that can be from any layer or it can be missing as well. I strongly believe that API response building should be API servers responsibility with respect to API Spec definition and it must not just pass through what comes from backbone services / database.

Reply:
Backbone's response model has 1 to 1 relationship with API server response model. API server merely acts as proxy to read the data from RDS in this case.

If something changes in the backbone, how would you ensure API server is not broken?